### ISSUE TITLE
Volume definitions in docker-compose.yml don't work out of the box with new `docker compose`

### ISSUE THREAD

**Issue by theseion** [coreruleset/coreruleset#2096]
### Describe the bug

Running `docker compose up modsec2-apache` produces an error due to a "bad" bind mount definition.

### Steps to reproduce

1. clone the repository
2. `cd` into the repository
3. run `docker compose --project-directory tests up modsec2-apache` -> error
4. run `docker-compose --project-directory tests up modsec2-apache` -> works fine

### Expected behaviour

I realize that this is probably a bug in the docker CLI but people (i.e. me) will be moving to `docker compose` and away from `docker-compose`. I expect both forms to work (it took me about an hour to figure out why this was happening).

### Actual behaviour

The `docker compose` version fails, the `docker-compose` version does not and creates the missing directories for the volume definitions.

### Additional context

The issue is with the short hand volume syntax in the docker-compose file. The specification says that missing directories will be created but `docker compose` fails to do that and, thus, the subsequent `run` command fails.

### Your Environment

* Operating System and version: macOS 11.3
* Docker Desktop version: 20.10.6, build 370c289
* Docker Compose version: version 1.29.1, build c34c88b2

**Comment by azurit** [comment]
Thanks for reporting this!

Maybe I misunderstood you, so correct me if i'm wrong: Isn't this a bug in `docker compose`?

**Comment by theseion** [comment]
Yes, you are correct, this is a bug in `docker compose`. However, I don't know when this will be fixed and I suspect many people will be using `docker compose` regardless and then run into this issue. I've opened this issue in the hopes that we can change the setup so that both invocations can work to make it easier for new contributors to get started.

**Comment by fzipi** [comment]
@theseion Works perfect for me (macOS 11.3.1):
```
❯ docker --version
Docker version 20.10.6, build 370c289
❯ docker compose -f tests/docker-compose.yml up -d modsec2-apache
[+] Running 3/3
 ⠿ Network tests_default      Created                                                                                                                                        5.4s
 ⠿ Container tests_backend_1  Started                                                                                                                                        4.3s
 ⠿ Container modsec2-apache   Started
```

Using `--project-directory`:
```
❯ docker compose --project-directory tests up modsec2-apache
[+] Running 3/3
 ⠿ Network tests_default      Created                                                                                                                                        3.5s
 ⠿ Container tests_backend_1  Created                                                                                                                                        0.4s
 ⠿ Container modsec2-apache   Created                                                                                                                                        0.5s
Attaching to modsec2-apache
modsec2-apache  | 172.20.0.1 - - [22/May/2021:11:15:46 +0000] "GET / HTTP/1.1" 200 9593
```

¯\_(ツ)_/¯

**Comment by theseion** [comment]
Well, it doesn't for me :) Did you use a fresh clone? Because the command will work when the directories for the mount exist (the `logs/*` directories specifically).

**Comment by fzipi** [comment]
That is a good explanation. Will take a second look.

**Comment by fzipi** [comment]
With that one, seems to fail:
```bash
❯ git clone git@github.com:coreruleset/coreruleset.git
Cloning into 'coreruleset'...
remote: Enumerating objects: 12579, done.
remote: Counting objects: 100% (278/278), done.
remote: Compressing objects: 100% (121/121), done.
remote: Total 12579 (delta 178), reused 230 (delta 154), pack-reused 12301
Receiving objects: 100% (12579/12579), 2.83 MiB | 828.00 KiB/s, done.
Resolving deltas: 100% (9429/9429), done.
❯ cd coreruleset
❯ docker compose --project-directory tests up modsec2-apache
[+] Running 1/2
 ⠿ Container tests_backend_1  Created                                                                                                                                        0.0s
 ⠹ Container modsec2-apache   Recreate                                                                                                                                       0.2s
Error response from daemon: invalid mount config for type "bind": bind source path does not exist: /tmp/coreruleset/tests/logs/modsec2-apache
❯ docker compose -f tests/docker-compose.yml up modsec2-apache
[+] Running 1/2
 ⠿ Container tests_backend_1                           Created                                                                                                               0.0s
 ⠹ Container c5a8d5b703ed_c5a8d5b703ed_modsec2-apache  Recreate                                                                                                              0.2s
Error response from daemon: invalid mount config for type "bind": bind source path does not exist: /tmp/coreruleset/tests/logs/modsec2-apache
```

**Comment by theseion** [comment]
Yes, that's the one. 👍

**Comment by fzipi** [comment]
We can go two different paths here:
- add an empty `logs/` directory, only to workaround this temporary failure.
- document it, and move forward until docker fixes this

**Comment by theseion** [comment]
We'd have to add all of `logs/modsec2-apache/apache2` and `logs/modsec3-nginx/nginx` but yes, that would be the simplest solution. When we do this, we could also make an adaptation to change the log paths to `/var/log/modsec2-apache` and `/var/log/modsec3-nginx`. That would remove the unnecessary extra directory layer and we would only have to create `logs/modsec2-apache` and `logs/modsec3-nginx`.

Documenting it would be fine, but we'll probably end up with people asking to fix this again because they haven't read the documentation ;).

**Comment by fzipi360** [comment]
The problem with changing it like that is that you need a different directory for storing the `modsec_audit.log` file. That's why I've created the additional directory.
```bash
❯ tree
.
├── modsec2-apache
│   ├── apache2
│   │   └── error.log
│   └── modsec_audit.log
└── modsec3-nginx
    ├── modsec_audit.log
    └── nginx
        ├── access.log
        └── error.log
```

**Comment by theseion** [comment]
I see. It works for apache but the NGiNX version used in the image has been compiled with that log directory hard coded. There are ways around that but they aren't really worth the effort. Well, it would have been nice but isn't too bad.

In that case let's create those directories?

**Comment by fzipi** [comment]
Sure.

If want to create also an issue in the modsecurity-docker for the nginx log directory, please go ahead!

**Comment by theseion** [comment]
I've opened #2103. Note that I moved the logs up one directory anyway because I think it is annoying. Instead, I opted to create the `nginx` directory, which will be populated with an `error.log` file by NGiNX but which we can simply ignore. I can understand if you don't like this setup. Let me know.

In addition, I added README files to the empty directories. Git will not (easily) store empty directories, so I've taken the opportunity to document why the directories are empty.
